enclave-runtime/src/rpc/author/mod.rs

```diff
@@ -161,25 +180,38 @@ where
 		let best_block_hash = Default::default();
 
 		match submission_mode {
-			TopSubmissionMode::Submit => Box::pin(
-				self.pool
-					.submit_one(
-						&generic::BlockId::hash(best_block_hash),
-						TX_SOURCE,
-						stf_operation,
-						shard,
+			TopSubmissionMode::Submit => self.execute_on_top_pool(
+				|t: &TopPoolGetter::TrustedOperationPool| -> PoolFuture<TxHash<TopPoolGetter::TrustedOperationPool>, RpcError> {
+					Box::pin(
+						t.submit_one(
+							&generic::BlockId::hash(best_block_hash),
+							TX_SOURCE,
+							stf_operation,
+							shard,
+						)
+						.map_err(map_top_error::<TopPoolGetter::TrustedOperationPool>),
 					)
-					.map_err(map_top_error::<P>),
+				},
+				|e| -> PoolFuture<TxHash<TopPoolGetter::TrustedOperationPool>, RpcError> {
+						Box::pin(ready(Err(e.into())))
+					},
```

Comment:
This has gotten a bit more complex now - we call `self.execute_on_top_pool()` and provide the function/closure we want to execute, and also a function to map the mutex error.